File: packages/python/goldenflow/goldenflow/transforms/numeric.py
from __future__ import annotations

import math

from goldenflow._polars_lazy import pl
from goldenflow.transforms import register_transform
from goldenflow.transforms._native import (
    abs_value_native,
    clamp_native,
    comma_decimal_native,
    currency_strip_native,
    fill_zero_native,
    fraction_to_decimal_native,
    ordinal_to_int_native,
    percentage_normalize_native,
    roman_to_int_native,
    round_native,
    scientific_to_decimal_native,
    to_integer_native,
)

# Pure-Python reference for goldenflow-core's ``numeric`` kernel. MUST
# reproduce the Rust kernel VALUE-for-VALUE (asserted by
# tests/transforms/test_identifiers_parity.py / test_numeric_kernels.py over
# tests/parity/identifiers_corpus.jsonl). This family outputs floats/ints, so
# parity is by VALUE, not string repr.


def _currency_strip_py(val: str | None) -> float | None:
    if val is None:
        return None
    filtered = "".join(c for c in str(val) if c.isdigit() or c in ".-")
    try:
        return float(filtered)
    except ValueError:
        return None


def _percentage_normalize_py(val: str | None) -> float | None:
    if val is None:
        return None
    v = str(val).strip()
    v = v.rstrip("%")
    v = v.strip()
    try:
        return float(v) / 100.0
    except ValueError:
        return None


def _to_integer_py(val: str | None) -> int | None:
    if val is None:
        return None
    try:
        return int(float(str(val).strip()))
    except ValueError:
        return None


def _comma_decimal_py(val: str | None) -> float | None:
    if val is None:
        return None
    v = str(val).strip()
    if "," not in v:
        try:
            return float(v)
        except ValueError:
            return None
    v = v.replace(".", "").replace(",", ".")
    try:
        return float(v)
    except ValueError:
        return None


def _scientific_to_decimal_py(val: str | None) -> float | None:
    if val is None:
        return None
    try:
        return float(str(val).strip())
    except ValueError:
        return None


def _round_f64_py(x: float, n: int) -> float:
    """Round-half-away-from-zero at the n-th decimal, via multiply/round/
    divide -- the SAME formula as goldenflow-core's ``round_f64`` kernel.
    Deliberately NOT Python's builtin ``round()`` (round-half-to-even)."""
    factor = 10.0**n
    scaled = x * factor
    rounded = math.floor(scaled + 0.5) if scaled >= 0 else math.ceil(scaled - 0.5)
    return rounded / factor


def _clamp_f64_py(x: float, min_val: float, max_val: float) -> float:
    if x < min_val:
        return min_val
    if x > max_val:
        return max_val
    return x


def _abs_f64_py(x: float) -> float:
    return abs(x)


def _currency_strip_series(series: pl.Series) -> pl.Series:
    native = currency_strip_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Utf8, strict=False).map_elements(
        _currency_strip_py, return_dtype=pl.Float64
    )


def _percentage_normalize_series(series: pl.Series) -> pl.Series:
    native = percentage_normalize_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Utf8, strict=False).map_elements(
        _percentage_normalize_py, return_dtype=pl.Float64
    )


def _to_integer_series(series: pl.Series) -> pl.Series:
    native = to_integer_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Utf8, strict=False).map_elements(
        _to_integer_py, return_dtype=pl.Int64
    )


# --- W5 numeric breadth: roman / fraction / ordinal parsers -------------------

_ROMAN_VAL = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
_ROMAN_TABLE = [
    (1000, "M"), (900, "CM"), (500, "D"), (400, "CD"), (100, "C"),
    (90, "XC"), (50, "L"), (40, "XL"), (10, "X"), (9, "IX"),
    (5, "V"), (4, "IV"), (1, "I"),
]


def _int_to_roman(n: int) -> str:
    out = []
    for v, sym in _ROMAN_TABLE:
        while n >= v:
            out.append(sym)
            n -= v
    return "".join(out)


def _roman_to_int_py(val: str | None) -> int | None:
    if val is None:
        return None
    t = str(val).strip().upper()
    if not t:
        return None
    chars = list(t)
    total = 0
    for i, c in enumerate(chars):
        cur = _ROMAN_VAL.get(c)
        if cur is None:
            return None
        nxt = _ROMAN_VAL.get(chars[i + 1], 0) if i + 1 < len(chars) else 0
        total += -cur if cur < nxt else cur
    if not 1 <= total <= 3999:
        return None
    return total if _int_to_roman(total) == t else None


def _ordinal_suffix(n: int) -> str:
    if 11 <= n % 100 <= 13:
        return "th"
    return {1: "st", 2: "nd", 3: "rd"}.get(n % 10, "th")


def _ordinal_to_int_py(val: str | None) -> int | None:
    if val is None:
        return None
    t = str(val).strip().lower()
    digits = ""
    for c in t:
        if c.isascii() and c.isdigit():
            digits += c
        else:
            break
    if not digits:
        return None
    n = int(digits)
    return n if t[len(digits):] == _ordinal_suffix(n) else None


def _parse_fraction(s: str) -> float | None:
    if "/" not in s:
        return None
    num_s, den_s = s.split("/", 1)
    try:
        num = float(num_s.strip())
        den = float(den_s.strip())
    except ValueError:
        return None
    if den == 0.0:
        return None
    return num / den


def _fraction_to_decimal_py(val: str | None) -> float | None:
    if val is None:
        return None
    t = str(val).strip()
    if not t:
        return None
    parts = t.split(None, 1)
    if len(parts) == 2:
        frac_s = parts[1].strip()
        if "/" in frac_s:
            try:
                whole = int(parts[0].strip())
            except ValueError:
                return None
            frac = _parse_fraction(frac_s)
            if frac is None:
                return None
            return (whole - frac) if whole < 0 else (whole + frac)
        return None
    if "/" in t:
        return _parse_fraction(t)
    try:
        return float(t)
    except ValueError:
        return None


def _roman_to_int_series(series: pl.Series) -> pl.Series:
    native = roman_to_int_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Utf8, strict=False).map_elements(
        _roman_to_int_py, return_dtype=pl.Int64
    )


def _ordinal_to_int_series(series: pl.Series) -> pl.Series:
    native = ordinal_to_int_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Utf8, strict=False).map_elements(
        _ordinal_to_int_py, return_dtype=pl.Int64
    )


def _fraction_to_decimal_series(series: pl.Series) -> pl.Series:
    native = fraction_to_decimal_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Utf8, strict=False).map_elements(
        _fraction_to_decimal_py, return_dtype=pl.Float64
    )


@register_transform(
    name="currency_strip", input_types=["string", "numeric"], auto_apply=False, priority=50, mode="expr"
)
def currency_strip(column: str) -> pl.Expr:
    """Strip currency symbols and thousand separators, return numeric.

    Native-first (goldenflow-core's ``numeric::currency_strip`` kernel),
    dispatched via ``map_batches`` so the transform keeps its original
    ``expr``-mode signature; the pure-Python fallback is the value-exact
    reference this kernel replicates.
    """
    return pl.col(column).map_batches(_currency_strip_series, return_dtype=pl.Float64)


@register_transform(
    name="percentage_normalize",
    input_types=["string", "numeric"],
    auto_apply=False,
    priority=50,
    mode="expr",
)
def percentage_normalize(column: str) -> pl.Expr:
    """Strip trailing %, parse to float, divide by 100.

    Native-first (goldenflow-core's ``numeric::percentage_normalize``
    kernel), dispatched via ``map_batches``; the pure-Python fallback is the
    value-exact reference this kernel replicates.
    """
    return pl.col(column).map_batches(_percentage_normalize_series, return_dtype=pl.Float64)


@register_transform(
    name="round", input_types=["numeric"], auto_apply=False, priority=40, mode="series"
)
def round_values(series: pl.Series, n: int = 2) -> pl.Series:
    """Round to n decimal places (round-half-away-from-zero, see the
    goldenflow-core ``round_f64`` kernel doc for why this deliberately isn't
    the language builtin ``round()``).

    Native-first; the pure-Python fallback below is the value-exact
    reference this kernel replicates.
    """
    native = round_native(n)
    if native is not None:
        return native(series)
    return series.cast(pl.Float64, strict=False).map_elements(
        lambda x: None if x is None else _round_f64_py(x, n), return_dtype=pl.Float64
    )


@register_transform(
    name="clamp", input_types=["numeric"], auto_apply=False, priority=40, mode="series"
)
def clamp(series: pl.Series, min_val: float = 0.0, max_val: float = 1.0) -> pl.Series:
    """Clip values into [min_val, max_val].

    Native-first (goldenflow-core's ``numeric::clamp_f64`` kernel); the
    pure-Python fallback below is the value-exact reference this kernel
    replicates.
    """
    native = clamp_native(min_val, max_val)
    if native is not None:
        return native(series)
    return series.cast(pl.Float64, strict=False).map_elements(
        lambda x: None if x is None else _clamp_f64_py(x, min_val, max_val),
        return_dtype=pl.Float64,
    )


@register_transform(
    name="roman_to_int", input_types=["string"], auto_apply=False, priority=40, mode="expr"
)
def roman_to_int(column: str) -> pl.Expr:
    """Parse a Roman numeral to its integer (canonical forms only, 1..=3999).
    Native-first over goldenflow-core."""
    return pl.col(column).map_batches(_roman_to_int_series, return_dtype=pl.Int64)


@register_transform(
    name="ordinal_to_int", input_types=["string"], auto_apply=False, priority=40, mode="expr"
)
def ordinal_to_int(column: str) -> pl.Expr:
    """Parse an English ordinal (1st/2nd/3rd) to its integer. Native-first."""
    return pl.col(column).map_batches(_ordinal_to_int_series, return_dtype=pl.Int64)


@register_transform(
    name="fraction_to_decimal", input_types=["string", "numeric"], auto_apply=False, priority=40, mode="expr"
)
def fraction_to_decimal(column: str) -> pl.Expr:
    """Parse a fraction or mixed number (1/2, 3 3/4) to a float. Native-first."""
    return pl.col(column).map_batches(_fraction_to_decimal_series, return_dtype=pl.Float64)


@register_transform(
    name="to_integer",
    input_types=["string", "numeric"],
    auto_apply=False,
    priority=45,
    mode="expr",
)
def to_integer(column: str) -> pl.Expr:
    """Parse string to integer, truncating any decimal part.

    Native-first (goldenflow-core's ``numeric::to_integer`` kernel),
    dispatched via ``map_batches``; the pure-Python fallback is the
    value-exact reference this kernel replicates.
    """
    return pl.col(column).map_batches(_to_integer_series, return_dtype=pl.Int64)


@register_transform(
    name="abs_value",
    input_types=["numeric"],
    auto_apply=False,
    priority=40,
    mode="series",
)
def abs_value(series: pl.Series) -> pl.Series:
    """Return the absolute value.

    Native-first (goldenflow-core's ``numeric::abs_f64`` kernel); the
    pure-Python fallback below is the value-exact reference this kernel
    replicates.
    """
    native = abs_value_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Float64, strict=False).map_elements(
        lambda x: None if x is None else _abs_f64_py(x), return_dtype=pl.Float64
    )


@register_transform(
    name="fill_zero",
    input_types=["numeric"],
    auto_apply=False,
    priority=35,
    mode="series",
)
def fill_zero(series: pl.Series) -> pl.Series:
    """Replace null values with 0.

    Native-first (goldenflow-core's ``numeric::fill_zero`` kernel); the
    pure-Python fallback below is the value-exact reference this kernel
    replicates.
    """
    native = fill_zero_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Float64, strict=False).fill_null(0.0)


@register_transform(
    name="comma_decimal",
    input_types=["string", "numeric"],
    auto_apply=False,
    priority=48,
    mode="series",
)
def comma_decimal(series: pl.Series) -> pl.Series:
    """Convert European decimal format (1.234,56) to float (1234.56).

    Native-first (goldenflow-core's ``numeric::comma_decimal`` kernel); the
    pure-Python fallback below is the value-exact reference this kernel
    replicates.
    """
    native = comma_decimal_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Utf8, strict=False).map_elements(
        _comma_decimal_py, return_dtype=pl.Float64
    )


@register_transform(
    name="scientific_to_decimal",
    input_types=["string", "numeric"],
    auto_apply=False,
    priority=45,
    mode="series",
)
def scientific_to_decimal(series: pl.Series) -> pl.Series:
    """Convert scientific notation (1.5e3) to decimal (1500.0).

    Native-first (goldenflow-core's ``numeric::scientific_to_decimal``
    kernel); the pure-Python fallback below is the value-exact reference
    this kernel replicates.
    """
    native = scientific_to_decimal_native()
    if native is not None:
        return native(series)
    return series.cast(pl.Utf8, strict=False).map_elements(
        _scientific_to_decimal_py, return_dtype=pl.Float64
    )

File: packages/rust/extensions/goldenflow-core/src/numeric.rs
//! Owned numeric kernels (pyo3-free): string->number parsers (currency,
//! percentage, integer-truncation, EU comma-decimal, scientific notation) and
//! numeric-array ops (round, clamp, abs, fill-null-with-zero). These are the
//! reference implementations; the Python/TS fallbacks must reproduce their
//! VALUES exactly (byte/value-parity harness -- numeric comparison, not
//! string repr, since this family outputs floats/ints).
//!
//! Deliberately NOT using a `regex` crate dependency (mirrors the other
//! goldenflow-core kernels' no-regex policy) -- `currency_strip` hand-filters
//! chars instead of `[^\d.\-]` regex-replace.

/// Strip everything except ASCII digits, `.`, and `-`, then parse as `f64`.
/// `None` on parse failure (mirrors Polars `cast(Float64, strict=False)`
/// null-on-failure semantics).
pub fn currency_strip(s: &str) -> Option<f64> {
    let filtered: String = s
        .chars()
        .filter(|c| c.is_ascii_digit() || *c == '.' || *c == '-')
        .collect();
    filtered.parse::<f64>().ok()
}

/// Trim whitespace, strip trailing `%` character(s), trim again (in case
/// removing `%` exposes more whitespace), parse as `f64`, and divide by 100.
/// `None` on parse failure.
pub fn percentage_normalize(s: &str) -> Option<f64> {
    let trimmed = s.trim();
    let stripped = trimmed.trim_end_matches('%');
    let stripped = stripped.trim();
    stripped.parse::<f64>().ok().map(|v| v / 100.0)
}

/// Parse a string to `f64` then truncate toward zero to `i64` (mirrors the
/// old `int(float(val))` semantics / Polars `cast(Float64).cast(Int64)`,
/// both strict=False). `None` on parse failure.
pub fn to_integer(s: &str) -> Option<i64> {
    s.trim().parse::<f64>().ok().map(|v| v as i64)
}

/// Convert a European decimal format (`1.234,56`) to a plain `f64`
/// (`1234.56`). If the (trimmed) input has no comma, parse as-is (US format
/// or a plain number). `None` on parse failure.
pub fn comma_decimal(s: &str) -> Option<f64> {
    let trimmed = s.trim();
    if !trimmed.contains(',') {
        return trimmed.parse::<f64>().ok();
    }
    let converted = trimmed.replace('.', "").replace(',', ".");
    converted.parse::<f64>().ok()
}

/// Convert scientific notation (`1.5e3`) to a plain `f64` (`1500.0`). Trims
/// whitespace first; `None` on parse failure. Rust's `f64::from_str` already
/// accepts scientific notation, so this is just a trimmed parse.
pub fn scientific_to_decimal(s: &str) -> Option<f64> {
    s.trim().parse::<f64>().ok()
}

/// Round-half-away-from-zero at the `n`-th decimal place, computed via
/// multiply/round/divide (the KERNEL is the source of truth for this
/// rounding rule -- Python's fallback and TS's fallback must replicate this
/// exact formula, not their language's built-in `round()`, since e.g.
/// Python's `round()` uses round-half-to-even). `n` may be negative (rounds
/// to the left of the decimal point).
pub fn round_f64(x: f64, n: i32) -> f64 {
    let factor = 10f64.powi(n);
    let scaled = x * factor;
    let rounded = if scaled >= 0.0 {
        (scaled + 0.5).floor()
    } else {
        (scaled - 0.5).ceil()
    };
    rounded / factor
}

/// Clamp `x` into `[min_val, max_val]`.
pub fn clamp_f64(x: f64, min_val: f64, max_val: f64) -> f64 {
    if x < min_val {
        min_val
    } else if x > max_val {
        max_val
    } else {
        x
    }
}

/// Absolute value.
pub fn abs_f64(x: f64) -> f64 {
    x.abs()
}

/// Replace a null value with `0.0`; a present value passes through
/// unchanged. Operates on `Option<f64>` since this is fundamentally a
/// null-handling op, not a value transform.
pub fn fill_zero(x: Option<f64>) -> f64 {
    x.unwrap_or(0.0)
}

/// The canonical Roman-numeral suffix would be complex to validate loosely, so
/// this parses to an integer then round-trips: a value only validates if it
/// re-encodes to exactly the (uppercased, trimmed) input. Range 1..=3999.
/// `None` for empty / non-Roman chars / malformed forms (e.g. `IIII`).
pub fn roman_to_int(s: &str) -> Option<i64> {
    let t: String = s.trim().to_uppercase();
    if t.is_empty() {
        return None;
    }
    let val = |c: char| -> Option<i64> {
        Some(match c {
            'I' => 1,
            'V' => 5,
            'X' => 10,
            'L' => 50,
            'C' => 100,
            'D' => 500,
            'M' => 1000,
            _ => return None,
        })
    };
    let chars: Vec<char> = t.chars().collect();
    let mut total = 0i64;
    for i in 0..chars.len() {
        let cur = val(chars[i])?;
        let nxt = if i + 1 < chars.len() {
            val(chars[i + 1])?
        } else {
            0
        };
        if cur < nxt {
            total -= cur;
        } else {
            total += cur;
        }
    }
    if !(1..=3999).contains(&total) {
        return None;
    }
    // Round-trip validation rejects non-canonical forms (IIII, VX, ...).
    if int_to_roman(total) == t {
        Some(total)
    } else {
        None
    }
}

/// Encode 1..=3999 as a canonical Roman numeral (helper for `roman_to_int`'s
/// round-trip check).
fn int_to_roman(mut n: i64) -> String {
    const TABLE: &[(i64, &str)] = &[
        (1000, "M"),
        (900, "CM"),
        (500, "D"),
        (400, "CD"),
        (100, "C"),
        (90, "XC"),
        (50, "L"),
        (40, "XL"),
        (10, "X"),
        (9, "IX"),
        (5, "V"),
        (4, "IV"),
        (1, "I"),
    ];
    let mut out = String::new();
    for &(v, sym) in TABLE {
        while n >= v {
            out.push_str(sym);
            n -= v;
        }
    }
    out
}

/// Parse a fraction or mixed number to `f64`: `1/2` -> 0.5, `3 3/4` -> 3.75,
/// `-1/2` -> -0.5, a plain number (`5`, `2.5`) -> itself. `None` on parse
/// failure or division by zero.
pub fn fraction_to_decimal(s: &str) -> Option<f64> {
    let t = s.trim();
    if t.is_empty() {
        return None;
    }
    // Mixed number: "<whole> <num>/<den>".
    if let Some((whole_str, frac_str)) = t.split_once(char::is_whitespace) {
        let frac_str = frac_str.trim();
        if frac_str.contains('/') {
            let whole: i64 = whole_str.trim().parse().ok()?;
            let frac = parse_fraction(frac_str)?;
            return Some(if whole < 0 {
                whole as f64 - frac
            } else {
                whole as f64 + frac
            });
        }
        return None; // whitespace but not a mixed fraction -> not parseable
    }
    if t.contains('/') {
        return parse_fraction(t);
    }
    t.parse::<f64>().ok()
}

/// Parse a bare `num/den` fraction to `f64`. `None` on parse failure or a zero
/// denominator.
fn parse_fraction(s: &str) -> Option<f64> {
    let (num, den) = s.split_once('/')?;
    let num: f64 = num.trim().parse().ok()?;
    let den: f64 = den.trim().parse().ok()?;
    if den == 0.0 {
        return None;
    }
    Some(num / den)
}

/// Parse an English ordinal (`1st`, `22nd`, `3RD`, `11th`) to its integer.
/// The suffix must be the CORRECT one for the number (so `1th` -> `None`). A
/// bare number with no suffix is NOT an ordinal (`5` -> `None`). `None` for
/// empty / malformed input.
pub fn ordinal_to_int(s: &str) -> Option<i64> {
    let t = s.trim().to_lowercase();
    let digits: String = t.chars().take_while(char::is_ascii_digit).collect();
    if digits.is_empty() {
        return None;
    }
    let n: i64 = digits.parse().ok()?;
    let suffix = &t[digits.len()..];
    if suffix == ordinal_suffix(n) {
        Some(n)
    } else {
        None
    }
}

/// The English ordinal suffix for `n` (`st`/`nd`/`rd`/`th`).
fn ordinal_suffix(n: i64) -> &'static str {
    let m = n % 100;
    if (11..=13).contains(&m) {
        return "th";
    }
    match n % 10 {
        1 => "st",
        2 => "nd",
        3 => "rd",
        _ => "th",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn currency_strip_basic() {
        assert_eq!(currency_strip("$1,234.56"), Some(1234.56));
        assert_eq!(currency_strip("-$42.00"), Some(-42.0));
        assert_eq!(currency_strip("USD 100"), Some(100.0));
        assert_eq!(currency_strip(""), None);
        assert_eq!(currency_strip("abc"), None);
    }

    #[test]
    fn percentage_normalize_basic() {
        assert_eq!(percentage_normalize("50%"), Some(0.5));
        assert_eq!(percentage_normalize(" 12.5 % "), Some(0.125));
        assert_eq!(percentage_normalize("100"), Some(1.0));
        assert_eq!(percentage_normalize("abc%"), None);
        assert_eq!(percentage_normalize(""), None);
    }

    #[test]
    fn to_integer_basic() {
        assert_eq!(to_integer("42"), Some(42));
        assert_eq!(to_integer("42.9"), Some(42));
        assert_eq!(to_integer("-3.5"), Some(-3));
        assert_eq!(to_integer("abc"), None);
        assert_eq!(to_integer(""), None);
    }

    #[test]
    fn comma_decimal_basic() {
        assert_eq!(comma_decimal("1.234,56"), Some(1234.56));
        assert_eq!(comma_decimal("1234.56"), Some(1234.56));
        assert_eq!(comma_decimal("42"), Some(42.0));
        assert_eq!(comma_decimal("abc,de"), None);
        assert_eq!(comma_decimal(""), None);
    }

    #[test]
    fn scientific_to_decimal_basic() {
        assert_eq!(scientific_to_decimal("1.5e3"), Some(1500.0));
        assert_eq!(scientific_to_decimal(" 2E-2 "), Some(0.02));
        assert_eq!(scientific_to_decimal("abc"), None);
        assert_eq!(scientific_to_decimal(""), None);
    }

    #[test]
    fn round_f64_basic() {
        assert_eq!(round_f64(2.345, 2), 2.35);
        // 2.005 is not exactly representable; its nearest f64 is slightly
        // ABOVE 2.005, so *100 rounds up (a real float-representation
        // quirk, not a bug -- Python's `float("2.005")` has the same value).
        assert_eq!(round_f64(2.005, 2), 2.01);
        assert_eq!(round_f64(-2.345, 2), -2.35);
        assert_eq!(round_f64(1234.0, -2), 1200.0);
    }

    #[test]
    fn clamp_f64_basic() {
        assert_eq!(clamp_f64(0.5, 0.0, 1.0), 0.5);
        assert_eq!(clamp_f64(-0.5, 0.0, 1.0), 0.0);
        assert_eq!(clamp_f64(1.5, 0.0, 1.0), 1.0);
    }

    #[test]
    fn abs_f64_basic() {
        assert_eq!(abs_f64(-3.5), 3.5);
        assert_eq!(abs_f64(3.5), 3.5);
    }

    #[test]
    fn fill_zero_basic() {
        assert_eq!(fill_zero(None), 0.0);
        assert_eq!(fill_zero(Some(5.0)), 5.0);
    }
}
